README
NAME
    WWW::3172::SearchEngine - An interface to a search engine index of pages
    crawled by WWW::3172::Crawler

VERSION
    version 0.001

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/WWW-3172-SearchEngine/>.

    The development version lives at
    <http://github.com/doherty/WWW-3172-SearchEngine> and may be cloned from
    <git://github.com/doherty/WWW-3172-SearchEngine.git>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

SOURCE
    The development version is on github at
    <http://github.com/doherty/WWW-3172-SearchEngine> and may be cloned from
    <git://github.com/doherty/WWW-3172-SearchEngine.git>

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <https://github.com/doherty/WWW-3172-SearchEngine/issues>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module {{$dist->name}}

{{$NEXT}}

0.001     2011-10-19
 - First version


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "An interface to a search engine index of pages crawled by WWW::3172::Crawler",
  "dist_author" => [
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "WWW-3172-SearchEngine",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "WWW::3172::SearchEngine",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dancer" => 0,
    "Dancer::Plugin::Database" => 0,
    "HTML::Entities" => 0,
    "Lingua::Stem::Snowball" => 0,
    "Lingua::StopWords" => 0,
    "List::MoreUtils" => 0,
    "Text::Unidecode" => 0,
    "WWW::Form" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/app.pl"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/WWW/3172/SearchEngine.pm
package WWW::3172::SearchEngine;
use strict;
use warnings;
use Dancer ':syntax';
use Dancer::Plugin::Database;
# ABSTRACT: An interface to a search engine index of pages crawled by WWW::3172::Crawler
# VERSION

use WWW::Form;
use HTML::Entities qw(encode_entities);
use List::MoreUtils qw(uniq);
use Lingua::Stem::Snowball;
use Lingua::StopWords qw(getStopWords);
use Text::Unidecode qw(unidecode);

my $stopwords = getStopWords('en');
my $stemmer   = Lingua::Stem::Snowball->new(
    lang    => 'en',
    encoding=> 'UTF-8',
);

get '/' => sub {
    my $form = WWW::Form->new(
        { 'q' => { label => 'Search query', type => 'text' } },
        { 'q' => '' },
        ['q']
    );

    template 'index', {
        form => $form->getFormHTML(
            action  => '/s',
            name    => 'search-form',
            submit_label => 'Search',
        ),
    };
};

any ['get', 'post'] => '/s' => sub {
    my $query = params->{q};

    my @words;
    {
        my $punct = quotemeta q{.,[]()<>{}+-=_'"\|};
        my @split = map { ## no critic (ControlStructures::ProhibitMutatingListFunctions)
            s{\A[$punct]}{};
            s{[$punct]\Z}{};

            length($_) < 2 || exists $stopwords->{$_}
                ? ()
                : unidecode($_)
        } split /\s+/, $query;
        @words = $stemmer->stem([uniq @split]);
    }
    redirect '/' unless @words;

    my $sql = sprintf(<<'SQL', (('?,' x (scalar(@words) - 1)) . '?' ) );
SELECT url_str FROM url WHERE url_id in (
    SELECT url_id FROM url_words WHERE word_id in (
        SELECT word_id FROM words WHERE word_str in (%s)
    ) ORDER BY url_word_count DESC
)
SQL

    my $sth = database->prepare($sql);
    $sth->execute(@words);

    my @urls;
    while (my $data = $sth->fetchrow_hashref) {
        push @urls, $data->{url_str};
    }

    return template 'results', {
        results => '<p>No results were found for <strong>' . encode_entities($query) . '</strong></p>'
    } unless @urls;

    my $html = '<ul>';
    foreach my $url (@urls) {
        my $safe_url = encode_entities($url);
        $html .= "<li><a href='$safe_url'>$safe_url</a></li>\n";
    }
    $html .= '</ul>';

    template 'results', { results => $html };
};

true;



Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "An interface to a search engine index of pages crawled by WWW::3172::Crawler",
  "AUTHOR" => "Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "WWW-3172-SearchEngine",
  "EXE_FILES" => [
    "bin/app.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "WWW::3172::SearchEngine",
  "PREREQ_PM" => {
    "Dancer" => 0,
    "Dancer::Plugin::Database" => 0,
    "HTML::Entities" => 0,
    "Lingua::Stem::Snowball" => 0,
    "Lingua::StopWords" => 0,
    "List::MoreUtils" => 0,
    "Text::Unidecode" => 0,
    "WWW::Form" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



